### PR TITLE
General update and switch to std::mutex

### PR DESCRIPTION
**Pull request by dzenanz**
I was getting some errors with this. An independent test/update of this PR would be welcome.

### CODE REVIEW

File: itk-module.cmake [hunk 1]
```diff
@@ -21,7 +21,16 @@ itk_module(VariationalRegistration
...
+    ITKIOGDCM
+    ITKIOMeta
+    ITKIOJPEG
+    ITKIOPNG
+    ITKIOTIFF
+    ITKIOBMP
+    ITKIOVTK
+    ITKIONRRD
+    ITKIOGIPL
+    ITKIONIFTI
```
Comment (jhlegarreta) [inline itk-module.cmake:33]: May be these could be replaced by `ITKImageIO`? Was also applied in [LesionSizingToolkit](https://github.com/InsightSoftwareConsortium/LesionSizingToolkit/pull/26) 😉

Comment (thewtex) [reply itk-module.cmake:33]: :+1:

Comment (dzenanz) [reply itk-module.cmake:33]: I tried that, I was still getting some errors. But those might have been unrelated errors.

**Comment by dzenanz** [comment]
As there are other pull requests, maybe we should wait until they are merged or closed before trying to fix this PR? Rebasing on top of other fixes might help.

**Comment by jhlegarreta** [comment]
Sounds like a good plan.

**Comment by dzenanz** [comment]
On my computer, `30% tests passed, 14 tests failed out of 20`. All 2D tests need an update to baseline images. 3D tests have some weird artifacts around the edges of images, so we shouldn't just take the current outputs as baselines.

File: itk-module.cmake (hunk 1 above)
Comment (dzenanz) [reply itk-module.cmake:33]: I tried it again, and again it did not work ☹️

Comment (jhlegarreta) [reply itk-module.cmake:33]: Well, may be we just need to put this aside for now, and may be at a latter time we might be able to discover the reason :pensive:. Thanks for trying Dženan.

**Comment by dzenanz** [comment]
Could we merge this as is? It is an improvement, and there are plenty of things to do to fix this completely.

**Comment by thewtex** [comment]
We should try to keep CI green, if possible. There are build errors:

```
In file included from /ITKVariationalRegistration/include/itkVariationalRegistrationDiffusionRegularizer.h:191:0,
                 from /ITKVariationalRegistration/include/itkVariationalRegistrationFilter.h:26,
                 from /ITKVariationalRegistration/test/VariationalRegistrationFilterTest.cxx:19:
/ITKVariationalRegistration/include/itkVariationalRegistrationDiffusionRegularizer.hxx: In static member function 'static void* itk::VariationalRegistrationDiffusionRegularizer<TDisplacementField>::CalcBufferCallback(void*)':
/ITKVariationalRegistration/include/itkVariationalRegistrationDiffusionRegularizer.hxx:270:10: error: 'ITK_THREAD_RETURN_DEFAULT_VALUE' was not declared in this scope
   return ITK_THREAD_RETURN_DEFAULT_VALUE;
```

**Comment by dzenanz** [comment]
That was due to a change Hans did not long ago. The CI infrastructure should be updated to use a newer commit. Do you want me to do it, or will you do it yourself @thewtex?

**Comment by thewtex** [comment]
> That was due to a change Hans did not long ago. The CI infrastructure should be updated to use a newer commit. Do you want me to do it, or will you do it yourself @thewtex?

@dzenanz please go ahead.

**Comment by dzenanz** [comment]
65% tests passed, 7 tests failed out of 20

Label Time Summary:
VariationalRegistration    = 241.04 sec*proc (20 tests)

Total Test time (real) = 155.20 sec

The following tests FAILED:
	  6 - VariationalRegistrationGaussian2DTest (Failed)
	 15 - VariationalRegistrationGaussian3DTest (Failed)
	 16 - VariationalRegistrationDiffusive3DTest (Failed)
	 17 - VariationalRegistrationSSD3DTest (Failed)
	 18 - VariationalRegistrationNCC3DTest (Failed)
	 19 - VariationalRegistrationDiffeomorph3DTest (Failed)
	 20 - VariationalRegistrationSymDiff3DTest (Failed)

**Comment by dzenanz** [comment]
Should we merge this as is? It is big improvement.

**Comment by hjmjohnson** [comment]
@dzenanz I'd suggest that this the failing test be commented out so that the dashboard is green and make an issue to fix the rest of the tests.

**Comment by dzenanz** [comment]
All the 3D tests are failing, I think. And one 2D test is crashing. I would rather leave that painfully obvious then temporarily sweep it under the carpet.

**Comment by thewtex** [comment]
Please keep `master` in a working state.

**Comment by dzenanz** [comment]
@thewtex Master is not in a working state, as can be seen on top of the [README.rst](https://github.com/InsightSoftwareConsortium/ITKVariationalRegistration#itkvariationalregistration). Master doesn't even compile, and this PR will bring it closer to working (`65% tests passed`).

File: include/itkVariationalRegistrationElasticRegularizer.hxx [hunk 2]
```diff
@@ -345,9 +345,9 @@ VariationalRegistrationElasticRegularizer< TDisplacementField >
 ::SolveElasticLESThreaderCallback( void * arg )
 {
   //Get MultiThreader struct
-  auto* threadStruct = (MultiThreader::ThreadInfoStruct *) arg;
-  int threadId = threadStruct->ThreadID;
-  int threadCount = threadStruct->NumberOfThreads;
+  auto* threadStruct = (MultiThreaderBase::WorkUnitInfo *) arg;
```
Comment (thewtex) [inline include/itkVariationalRegistrationElasticRegularizer.hxx:348]: If these are using the classing threading backend, should they be calling `this->DynamicMultiThreadingOff()`.

**Comment by thewtex** [comment]
> @thewtex Master is not in a working state, as can be seen on top of the README.rst. Master doesn't even compile, and this PR will bring it closer to working (65% tests passed).

Ok, then thanks for moving the builds forward. :+1: 

Please see inline comment regarding `this->DynamicMultiThreadingOff()`. If it does not apply, then please go ahead and merge.

File: include/itkVariationalRegistrationElasticRegularizer.hxx (hunk 2 above)
Comment (dzenanz) [reply include/itkVariationalRegistrationElasticRegularizer.hxx:348]: This class uses `GenerateData`, which calls `SolveElasticLES`, which sets up multi-threading.

**Pull request merged**